### PR TITLE
Fix liaison/grpc's leaked goroutines

### PR DESCRIPTION
**Pull request by mikechengwei**
Fixes https://github.com/apache/skywalking/issues/9463.

Signed-off-by: mikechengwei [842725815@qq.com](mailto:842725815@qq.com)

### CODE REVIEW

File: banyand/discovery/discovery.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ type ServiceRepo interface {
 	run.Unit
 	bus.Subscriber
 	bus.Publisher
+	GracefulStop()
```
Comment (hanahmily) [inline banyand/discovery/discovery.go:36]: Use `run.Service` in place of `GracefulStop`. The lifecycle framework will stop the repo gracefully.

File: banyand/liaison/grpc/registry_test.go [hunk 2]
```diff
@@ -193,12 +200,16 @@ func setupForRegistry() func() {
 		preloadStreamSvc,
 		tcp,
 	)
+	moduleDeferFunc := func() {
```
Comment (hanahmily) [inline banyand/liaison/grpc/registry_test.go:203]: Could you remove this defer function? If the repo uses the lifecycle framework to stop, `deferFunc` will help stop this repo. You don't have to add a separate one.

**Review by hanahmily — CHANGES_REQUESTED** [review]

File: banyand/discovery/discovery.go (hunk 1 above)
Comment (hanahmily) [reply banyand/discovery/discovery.go]: Add `stopCh` to ServiceRepo as https://github.com/apache/skywalking-banyandb/blob/main/banyand/queue/local.go#L48 did.

Comment (mikechengwei) [reply banyand/discovery/discovery.go]: ok

**Comment by mikechengwei** [comment]
@hanahmily please help me review it again ,thanks.

File: pkg/test/setup/setup.go [hunk 3]
```diff
@@ -91,7 +91,6 @@ func modules(flags []string) func() {
 
 	return test.SetupModules(
 		flags,
-		repo,
```
Comment (hanahmily) [inline pkg/test/setup/setup.go:94]: Why do you move `repo` to the last?

Comment (hanahmily) [reply pkg/test/setup/setup.go:94]: The order of sequence should be equal to https://github.com/apache/skywalking-banyandb/blob/main/banyand/internal/cmd/standalone.go#L81-L92

File: banyand/measure/measure_suite_test.go [hunk 4]
```diff
@@ -102,7 +105,6 @@ func setUp() (*services, func()) {
 	flags = append(flags, "--etcd-listen-client-url="+listenClientURL, "--etcd-listen-peer-url="+listenPeerURL)
 	moduleDeferFunc := test.SetupModules(
 		flags,
-		repo,
```
Comment (hanahmily) [inline banyand/measure/measure_suite_test.go:105]: If you remove `repo`, `moduleDeferFunc` can not stop it gracefully.

File: banyand/stream/stream_suite_test.go [hunk 5]
```diff
@@ -101,7 +104,6 @@ func setUp() (*services, func()) {
 	flags = append(flags, "--etcd-listen-client-url="+listenClientURL, "--etcd-listen-peer-url="+listenPeerURL)
 	moduleDeferFunc := test.SetupModules(
 		flags,
-		repo,
```
Comment (hanahmily) [inline banyand/stream/stream_suite_test.go:104]: You can't remove it.

**Review by hanahmily — CHANGES_REQUESTED** [review]

File: pkg/test/setup/setup.go (hunk 3 above)
Comment (mikechengwei) [reply pkg/test/setup/setup.go:94]: ok

File: banyand/measure/measure_suite_test.go (hunk 4 above)
Comment (mikechengwei) [reply banyand/measure/measure_suite_test.go:105]: If it is not removed, the unit test will block at https://github.com/apache/skywalking-banyandb/blob/1c19243df23f8350ea5c1542fd914c49e8698960/pkg/test/setup.go#L49 and cannot be terminated normally due to discovery_mock `gracefulstop` method.

log screensnapt:
image

Comment (mikechengwei) [reply banyand/measure/measure_suite_test.go:105]: Do you have any advice?

File: banyand/measure/measure_suite_test.go [hunk 6]
```diff
@@ -74,6 +74,9 @@ func setUp() (*services, func()) {
 	// Init Discovery
 	repo := discovery.NewMockServiceRepo(ctrl)
 	repo.EXPECT().NodeID().AnyTimes()
+	repo.EXPECT().Name().AnyTimes()
+	repo.EXPECT().Serve().AnyTimes()
+	repo.EXPECT().GracefulStop().AnyTimes()
```
Comment (hanahmily) [inline banyand/measure/measure_suite_test.go:79]:
```suggestion
	stopCh := make(chan struct{})
	repo.EXPECT().Serve().Return(stopCh).Times(1)
	repo.EXPECT().GracefulStop().Do(func() { close(stopCh) }).Times(1)
```

**Review by hanahmily — CHANGES_REQUESTED** [review]
You had better inject `stopCh` into the mocker.

File: banyand/measure/measure_suite_test.go (hunk 4 above)
Comment (hanahmily) [reply banyand/measure/measure_suite_test.go:105]: https://github.com/apache/skywalking-banyandb/pull/236#discussion_r1064253100

**Comment by mikechengwei** [comment]
I found that it is still a block, I will spend some time to troubleshoot.

**Comment by mikechengwei** [comment]
@hanahmily  please help me review it again, thanks for a lot .

**Review by hanahmily — APPROVED** [review]
LGTM.  Thank you

**Pull request merged**